File: week8/p7.c
#include <limits.h>
#include <stdbool.h>

struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

bool isBSTUtil(struct Node* node, long long minVal, long long maxVal) {
    if (node == NULL) return true;
    if (node->data <= minVal || node->data >= maxVal) return false;
    return isBSTUtil(node->left, minVal, node->data) &&
           isBSTUtil(node->right, node->data, maxVal);
}

bool isBST(struct Node* root) {
    return isBSTUtil(root, LLONG_MIN, LLONG_MAX);
}

File: week8/p7.py
'''
class Node:
    def __init__(self, val):
        self.data = val
        self.left = None
        self.right = None
'''

class Solution:
    def isBST(self, root):
        def helper(node, low, high):
            if not node:
                return True
            if node.data <= low or node.data >= high:
                return False
            return helper(node.left, low, node.data) and helper(node.right, node.data, high)
        
        return helper(root, float('-inf'), float('inf'))
